packages/go_router/example/lib/stateful_shell_route.dart

```diff
@@ -0,0 +1,339 @@
+        },
+
+        // If it's necessary to customize the Page for StatefulShellRoute,
+        // provide a pageBuilder function instead of the builder, for example:
+        // pageBuilder: (BuildContext context, GoRouterState state,
+        //             StatefulNavigationShell navigationShell) {
+        //   return NoTransitionPage<dynamic>(
+        //       child: ScaffoldWithNavBar(navigationShell: navigationShell));
+        // },
+      ),
+    ],
+  );
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      routerConfig: _router,
+    );
+  }
+}
+
+/// Builds the "shell" for the app by building a Scaffold with a
+/// BottomNavigationBar, where [child] is placed in the body of the Scaffold.
+class ScaffoldWithNavBar extends StatelessWidget {
+  /// Constructs an [ScaffoldWithNavBar].
+  const ScaffoldWithNavBar({
+    required this.navigationShell,
+    Key? key,
+  }) : super(key: key ?? const ValueKey<String>('ScaffoldWithNavBar'));
+
+  /// The navigation shell and container for the branch Navigators.
+  final StatefulNavigationShell navigationShell;
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      body: navigationShell,
+      bottomNavigationBar: BottomNavigationBar(
+        // Here, the items of BottomNavigationBar are hard coded. In a real
+        // world scenario, the items would most likely be generated from the
+        // branches of the shell route, which can be fetched using
+        // `navigationShell.route.branches`.
+        items: const <BottomNavigationBarItem>[
+          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Section A'),
+          BottomNavigationBarItem(icon: Icon(Icons.work), label: 'Section B'),
+          BottomNavigationBarItem(icon: Icon(Icons.tab), label: 'Section C'),
+        ],
+        currentIndex: navigationShell.currentIndex,
+        onTap: (int index) => _onTap(context, index),
+      ),
+    );
+  }
+
+  /// Navigate to the current location of the branch at the provided index when
+  /// tapping an item in the BottomNavigationBar.
+  void _onTap(BuildContext context, int index) {
+    if (index != navigationShell.currentIndex) {
+      // When navigating to a new branch, it's recommended to use the goBranch
+      // method, as doing so makes sure the last navigation state of the
+      // Navigator for the branch is restored.
+      navigationShell.goBranch(index);
+    } else {
+      // If tapping the bar item for the branch that is already active, go to
+      // the initial location instead. The method
+      // effectiveInitialBranchLocation can be used to get the (implicitly or
+      // explicitly) configured initial location in a convenient way.
+      GoRouter.of(context)
+          .go(navigationShell.effectiveInitialBranchLocation(index));
```

Comment:
It would be better if the user didn't have to write this code. It's not clear from reading this example if `goBranch` is intended to be used when the user taps on the current index.

Instead, why not support:

```dart
navigationShell.goBranch(index, goToInitialLocationWhenIndexMatches: true)
```

(The name should be something more compact, but you get the idea)

Reply:
I think it's also reasonable to make this the default behavior (go to the initial location when the user calls GoBranch with the current index) and the user can opt-out if they want. It would align better with how ShellRoute works in the examples.

Reply:
Actually, `goBranch` used to have an extra parameter called `resetLocation`. We removed that to simplify the API, but I think perhaps it would be preferable to add that one again, and instead remove (hide) `effectiveInitialBranchLocation`. That would simplify the case that the sample code showcases, i.e. returning to the initial location when tapping the current bottom nav bar item, while at the same time adding flexibility to support other scenarios (goToInitialLocationWhenIndexMatches seems pretty specific and narrow). 

Since the whole point of `StatefulShellRoute` is to maintain state for a set of parallel navigation stacks, it doesn't really make sense that the default behaviour would be to reset the branch when calling `goBranch`. The expected behaviour would be to maintain state. 

Therefore, I recommend the signature of `goBranch` be updated to: 

```dart
void goBranch(int index, {bool resetLocation = false})
```

Reply:
That seems simpler to me, but I don't know the context for the previous discussion.

Reply:
I'll let @chunhtai comment on this, but I believe the motivation was to simplify the API and remove things that weren't absolutely necessary. However, given the alternatives and that there is a clear and common use case for this functionality, my recommendation is that we do as I outlined above.

Reply:
I'll add a commit with this so you can properly review it.

Reply:
I originally suggest removing the resetLocation because I want to simply the pr, and I think this feature can be added later. If i were to choose between resetLocation or effectiveInitialBranchLocation, I will choose resetLocation. but maybe name it to something like
void goBranch(int index, {bool initialLocation = false})